net/interfaces/interfaces_windows.go

```diff
@@ -7,76 +7,91 @@ package interfaces
 import (
 	"fmt"
 	"log"
+	"net"
 	"net/url"
-	"os/exec"
 	"syscall"
 	"unsafe"
 
-	"go4.org/mem"
 	"golang.org/x/sys/windows"
 	"golang.zx2c4.com/wireguard/windows/tunnel/winipcfg"
 	"inet.af/netaddr"
 	"tailscale.com/tsconst"
-	"tailscale.com/util/lineread"
 )
 
 func init() {
 	likelyHomeRouterIP = likelyHomeRouterIPWindows
 	getPAC = getPACWindows
 }
 
-/*
-Parse out 10.0.0.1 from:
-
-Z:\>route print -4
-===========================================================================
-Interface List
- 15...aa 15 48 ff 1c 72 ......Red Hat VirtIO Ethernet Adapter
-  5...........................Tailscale Tunnel
-  1...........................Software Loopback Interface 1
-===========================================================================
-
-IPv4 Route Table
-===========================================================================
-Active Routes:
-Network Destination        Netmask          Gateway       Interface  Metric
-          0.0.0.0          0.0.0.0         10.0.0.1       10.0.28.63      5
-         10.0.0.0      255.255.0.0         On-link        10.0.28.63    261
-       10.0.28.63  255.255.255.255         On-link        10.0.28.63    261
-        10.0.42.0    255.255.255.0   100.103.42.106   100.103.42.106      5
-     10.0.255.255  255.255.255.255         On-link        10.0.28.63    261
-   34.193.248.174  255.255.255.255   100.103.42.106   100.103.42.106      5
-
-*/
 func likelyHomeRouterIPWindows() (ret netaddr.IP, ok bool) {
-	cmd := exec.Command("route", "print", "-4")
-	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
-	stdout, err := cmd.StdoutPipe()
+	rs, err := winipcfg.GetIPForwardTable2(windows.AF_INET)
 	if err != nil {
+		// TODO: It would be nice to log this...
 		return
 	}
-	if err := cmd.Start(); err != nil {
-		return
+
+	// Effective metric is sum of interface metric and route metric offset
+	setEffectiveMetric := func(r *winipcfg.MibIPforwardRow2) bool {
+		iface, err := r.InterfaceLUID.IPInterface(windows.AF_INET)
+		if err != nil {
+			// TODO: It would be nice to log this...
+			return false
+		}
+		r.Metric += iface.Metric
+		return true
 	}
-	defer cmd.Wait()
 
-	var f []mem.RO
-	lineread.Reader(stdout, func(lineb []byte) error {
-		line := mem.B(lineb)
-		if !mem.Contains(line, mem.S("0.0.0.0")) {
-			return nil
+	unspec := net.IPv4(0, 0, 0, 0)
+	var route *winipcfg.MibIPforwardRow2
+	metricUpdated := false
+
+	for _, r := range rs {
+		if r.Loopback || r.DestinationPrefix.PrefixLength != 0 || !r.DestinationPrefix.Prefix.IP().Equal(unspec) {
+			// Not a default route, so skip
+			continue
+		}
+
+		ip, ok := netaddr.FromStdIP(r.NextHop.IP())
+		if !ok || ip.IsZero() {
+			// Not a valid gateway, so skip (won't happen though)
+			continue
+		}
+
+		if route == nil {
+			route = &r
```

Comment:
I don't think this does what you think it does.

The "r" variable on line 48 is the same "r" for each iteration of the loop, and it gets reassigned for each element in rs.

So on the first pass, this isn't "saving" `r` in `route`... it's remembering the address of the never-changing `r` and then on the second iteration, it gets reassigned.

Easy fix: put an `r := r` on the first line inside the body, on line 49.

Reply:
Ohhhhhhhhhhhhh... THANKS!